File: 0528-random-pick-with-weight/0528-random-pick-with-weight.java
class Solution {
    int[] prefixSum;
    Random random = new Random();

    public Solution(int[] w) {
        int[] arr = new int[w.length];
        int cur = 0;
        
        for (int i = 0; i < w.length; i++) {
            cur += w[i];
            arr[i] = cur;
        }

        prefixSum = arr;
    }
    
    public int pickIndex() {
        int picked = random.nextInt(prefixSum[prefixSum.length - 1]) + 1;
        int left = 0;
        int right = prefixSum.length - 1;

        while (left < right) {
            int mid = (left + right) / 2;

            if (picked > prefixSum[mid]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        return left;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(w);
 * int param_1 = obj.pickIndex();
 */

File: 0528-random-pick-with-weight/0528-random-pick-with-weight.kt
class Solution(w: IntArray) {
    private val prefixSum = IntArray(w.size)

    init {
        var cur = 0
        for (i in w.indices) {
            cur += w[i]
            prefixSum[i] = cur
        }
    }

    fun pickIndex(): Int {
        val picked = (1..prefixSum.last()).random()
        var left = 0
        var right = prefixSum.size - 1

        while (left < right) {
            val mid = (left + right) / 2
            if (picked > prefixSum[mid]) left = mid + 1
            else right = mid
        }

        return left
    }

}

/**
 * Your Solution object will be instantiated and called as such:
 * var obj = Solution(w)
 * var param_1 = obj.pickIndex()
 */